File: src/tools/tonescale.js
function createToneScaleOf(tones) {

    const getNoteBy = (index) => tones[index % tones.length]
    const indexOf = (selected) => tones.findIndex(note => note === selected)
    const aggregateNotes = (scale, index) => [...scale, getNoteBy(index)]
    const scale = (offset, step) => [...offset, offset[offset.length - 1] + step]
    const getScaleOffsetBy = (note, mode) => mode.reduce(scale, [indexOf(note)])
    const getNotesWith = (offset) => (result, _, index) => aggregateNotes(result, index + offset)

    return {
        getRangeStartingWith: (note) => tones.reduce(getNotesWith(indexOf(note)), []),
        getScaleOf: (note, mode) => getScaleOffsetBy(note, mode).reduce(aggregateNotes, [])
    }
}

module.exports = createToneScaleOf

File: src/main/kotlin/tonescale.kt
val tones = arrayOf("a", "b")

// THIS DOES NOT WORK.

fun getNoteBy(index: Int) = tones[index % tones.size]
fun indexOf(note: String) = tones.indexOf(note)
fun aggregateNotes(scale: List<String>, idx: Int) = scale + getNoteBy(idx)
fun scale(offset: List<String>, step: String) = offset + offset.last() + step
//fun getScaleOffsetBy(note: String, mode: List<List<String>>) =  mode.fold(indexOf(note)) {offset, step -> offset + offset.last() + step}
